### PR TITLE
query: Made storeset more reliable and make it log on changes.

### PR DESCRIPTION
**Pull request by Bplotka**
Changes:
- gRPC dial is non-blocking, so we need to make intial Info call to make sure connection is ok for gossip.
- duplicates were wrongly handled
- log on every storeset change for better debugability

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: pkg/query/storeset.go [hunk 1]
```diff
@@ -168,34 +169,43 @@ func (s *storeRef) close() {
...
-		// New store or was unhealthy and was removed in the past - create new one.
-		conn, err := grpc.DialContext(ctx, addr, s.dialOpts...)
+
+	var err error
+	if ok {
+		// Check existing store. Is it healthy? What are current metadata.
+		st.labels, st.minTime, st.maxTime, err = spec.Metadata(ctx, st.StoreClient)
 		if err != nil {
-			return nil, errors.Wrap(err, "dialing connection")
+			st.close()
```
[the remark this answers is not in the document]
Comment (domgreen) [reply pkg/query/storeset.go]: ahh okay its deleted by checking `stores` and `s.stores` might be worth adding a comment to make it known it will be removed.

File: pkg/query/storeset.go [hunk 2]
```diff
@@ -215,21 +225,23 @@ func (s *StoreSet) Update(ctx context.Context) {
...
 				return err
 			}
+
 			innerMtx.Lock()
-			if dupSt, ok := stores[addr]; ok {
-				level.Error(s.logger).Log("msg", "duplicated address in gossip or static store nodes.", "storeAPIAddr", addr)
-				dupSt.close()
+			defer innerMtx.Unlock()
+
+			if _, ok := stores[addr]; ok {
```
Comment (domgreen) [inline pkg/query/storeset.go]: can probably do this straight away rather than try to `s.updateStore` as we will already know if the `addr` is in `store`.

File: cmd/thanos/query.go [hunk 3]
```diff
@@ -252,9 +252,8 @@ func (s *gossipSpec) Addr() string {
 	return s.addr
 }
 
-// Metadata method for gossip store tries get current peer state. If nothing is found, it means that gossip assumed
-// this host is unhealthy in the meantime.
-func (s *gossipSpec) Metadata(_ context.Context, _ storepb.StoreClient) (labels []storepb.Label, mint int64, maxt int64, err error) {
+// Metadata method for gossip store tries get current peer state.
+func (s *gossipSpec) Metadata(ctx context.Context, client storepb.StoreClient) (labels []storepb.Label, mint int64, maxt int64, err error) {
```
[the remark this answers is not in the document]
Comment (stefan-improbable) [reply cmd/thanos/query.go]: +1, I like `_` better so that you can see better that they are unused.

File: pkg/query/storeset.go (hunk 2 above)
Comment (stefan-improbable) [reply pkg/query/storeset.go]: +1 the `updateStore` method should be called `specToStore` and it shouldn't deal with maps at all

File: pkg/query/storeset.go [hunk 4]
```diff
@@ -215,21 +225,23 @@ func (s *StoreSet) Update(ctx context.Context) {
 			if err != nil {
 				return err
 			}
+
```
Comment (stefan-improbable) [inline pkg/query/storeset.go]: The code is a bit hard to follow. What about something like:
```
func isUnhealtyh(ctx, s store, unhelathyStores syncmap[addr]struct{}{}) { ... }

// Remove all unhealthy stores
unhealthyStores := synncmap.Map{}
for _, s := range s.stores {
   go(s) { isUnhealthy(ctx, s, &unhealthyStores }
}
wait

for _, unhealthy := range unhealthyStores {
  delete s.stores[unhelathy]
}

// Add new specs to stores
specs  := s.storeSpecs()
newStores := syncmap.Map{}
for _, spec := range specs {
  go(spec) {
    st := storeFromSpec
    if st != nil { newStores[addr] = st }
  }
}
wait
```
and so on. Feel free to ping me on slack for more info.

**Comment by domgreen** [comment]
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/query/storeset.go, line 47 at r1](https://reviewable.io/reviews/improbable-eng/thanos/355#-LDWS-liW0JNFw2AaMfW-r1-47:-LDWSX6H4L2iYQz1RfNM:b5on7ix) ([raw file](https://github.com/improbable-eng/thanos/blob/4315d307610d7e1d6928b1bb0ef50dffcc978606/pkg/query/storeset.go#L47)):*
<details><summary><i>Previously, domgreen (Dominic Green) wrote…</i></summary><blockquote>

No longer `static store` ? can probably get rid of the whole `method for static store`
</blockquote></details>

testing - will delete

---


*Comments from [Reviewable](https://reviewable.io/reviews/improbable-eng/thanos/355)*

File: cmd/thanos/query.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bplotka) [reply cmd/thanos/query.go]: sorry, I added something, moved this to `ctx` and `client` than removed stuff so they are no longer needed. Thanks for catching this.

File: pkg/query/storeset.go (hunk 4 above)
Comment (Bplotka) [reply pkg/query/storeset.go]: will do

File: pkg/query/storeset.go (hunk 2 above)
Comment (Bplotka) [reply pkg/query/storeset.go]: true

File: pkg/query/storeset.go (hunk 4 above)
Comment (Bplotka) [reply pkg/query/storeset.go]: refactored `Update` method.

File: pkg/query/storeset.go [hunk 5]
```diff
@@ -167,111 +177,122 @@ func (s *storeRef) close() {
...
+			ctx, cancel := context.WithTimeout(ctx, s.gRPCInfoCallTimeout)
+			defer cancel()
+
+			st, ok := s.stores[addr]
+			if ok {
+				// Check existing store. Is it healthy? What are current metadata?
+				labels, minTime, maxTime, err := spec.Metadata(ctx, st.StoreClient)
+				if err != nil {
+					// Peer unhealthy. Do not include in healthy stores.
```
Comment (stefan-improbable) [inline pkg/query/storeset.go]: I think you shouldn't use `errgroup.Group` here. Using errgroup signals that you have multiple goroutines that should end when one of them errors (which actually doesn't happen in your case). Also `err := g.Wait(); ` will return and print only the first error.
Instead you should use regular `go func()` with explicit waitgroup and log all errors inside the goroutine.
